### DISCUSSION TITLE
Ports: Arduino MKR WIFI 1010

### DISCUSSION THREAD

**Discussion by keenanjohnson** [micropython/micropython#10656] [SAMD] [answered]
Hello all! This is a great project, so thanks for the hard work everyone!

I was curious if it was reasonable to create a port of micropython for the Arduino MKR WIFI 1010. If so, I'm happy to take on the challenge of making the new board and pins files and submitting the PR.

It seems as if it should be straightforward as the [MCU (SAMD21)](https://github.com/micropython/micropython/tree/master/ports/samd/boards) and Wifi chipset ([Nina-w102](https://github.com/micropython/micropython/commit/43079aaf860cbc19d54cec1eca9d5897297b22ac)) are already both supported by micropython.

Anyone have any other gotchas or reasons why this shouldn't work?

Store Link: https://store-usa.arduino.cc/products/arduino-mkr-wifi-1010

**Comment by keenanjohnson** [comment]
I started working through the pin assignments here in my fork: https://github.com/keenanjohnson/micropython/commit/33fdc048f2835918e96b8f4de68524cd94c8cb02

**Comment by robert-hh** [comment, marked as answer]
Adapting the Pins should be easy, and the firmware should work straight away. There is already support for the Nina UBLOX module in the RP2040 port. So it might be possible to port that. But the show stopper for WiFi is the lack of available flash space, which is already being fully used. The only way to gain more space is reducing the size of the file system or dropping some MicroPython modules. But even that would not gain sufficient space for a good WiFi support, not to mention the lack of RAM. Unfortunately the Arduino board does not provide additional flash on the board like the Adafruit boards do. And I do not know why Arduino uses the SAMD21G18 for a relatively expensive board instead of the SAMD51G19, which offers way more Flash & RAM.

**Comment by robert-hh** [comment]
A word on your pins.csv file: define every GPIO only once and do not add conflicting assignments, like the same Dx for different GIO pins. That will just create confusion. A good starting point are as well the GENERIC boards. Pre-built binaries are here: https://github.com/robert-hh/Shared-Stuff/tree/master/samd_firmware, and I added the generic board definitions. If you get a REPL prompt with these, than little adaption is required.

[GENERIC_SAMD21G18.zip](https://github.com/micropython/micropython/files/10608162/GENERIC_SAMD21G18.zip)

**Comment by keenanjohnson** [comment]
Thanks very much for the help! I didn't know about those generic binaries and I will certainly try that out. My pins file I linked above is certainly still a draft haha.

Hmmm I hadn't considered how much space the wifi drivers may take, so thanks for pointing that out!

**Comment by keenanjohnson** [comment]
Update: [Flashing the Adafruit bootloader as described here](http://forrestbao.blogspot.com/2020/05/installing-circuitpython-on-arduino-mkr.html) and then moving the generic binary linked above allowed me to successfully enter the REPL on the MKR 1010.

**Comment by robert-hh** [comment]
That sounds good. The only drawback with the generic build is the lack of board names for pins. So have to use the CPU names or numbers, like `"PA07"` or `7`, `PB03` or `35` (32+3), ...

**Comment by keenanjohnson** [comment]
Is there a guide somewhere for how to disable the modules to try and get the size down to a reasonable enough limit to configure the wifi module?

**Comment by robert-hh** [comment]
The modules are enabled or disabled in the repective mpconfigxxx.h, For the samd port, these are mpconfigboard.h, mpconfigmcu.h, mpconfigport.h and py/mpconfig.h, in that order. To see all configurable modules you have to look into py/mpconfig.h. But WiFi requires a loot of flash. While working on the port, I noted the size of some features.
```
async :  ~1.5 k
unicode: ~2 k
simple float: ~12k
math lib: ~8k
Uart TX buffer: +308 Bytes (plus RAM for the buffer)
RTC: ~700 (SAMD51) ~500 (SAMD21)
Signal: 600 bytes.
```
So you could drop the math lib (sin, cos, ...), gaining 8k. There is a dedicated flag in mpconfigmcu.h for that. Or you can shrink the size of the flash file system, which is 64 K at the moment. That can be done in boards/samd21x18a.ld. With the PR #10233 comes a simple flag to set the code size in the mpconfigboard.mk file per board. It's a pity that Arduino did not add an extra flash chip to their relatively expensive board.

**Reply by keenanjohnson** [reply]
Thanks very much! I'll take a look and see what can be done :)
